### ISSUE TITLE
Editing KongIngress does not trigger an immediate sync

### ISSUE THREAD

**Issue by dsem** [Kong/kubernetes-ingress-controller#474]
NOTE: GitHub issues are reserved for **bug reports only**.
For anything else, please join the conversation
in Kong Nation https://discuss.konghq.com/c/kubernetes.

---

### Summary

Editing a KongIngress resource does not trigger an immediate sync of the Kubernetes configuration with the Kong configuration.  Instead, the configuration is only propagated at the next regular sync interval (10 minutes, I believe).

**Kong Ingress controller version**

v0.6

**Kong or Kong Enterprise version**
Kong

### Kubernetes version

**paste `kubectl version` output**

Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.0", GitCommit:"e8462b5b5dc2584fdcd18e6bcfe9f1e4d970a529", GitTreeState:"clean", BuildDate:"2019-06-19T16:40:16Z", GoVersion:"go1.12.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13+", GitVersion:"v1.13.11-gke.14", GitCommit:"56d89863d1033f9668ddd6e1c1aea81cd846ef88", GitTreeState:"clean", BuildDate:"2019-11-07T19:12:22Z", GoVersion:"go1.12.11b4", Compiler:"gc", Platform:"linux/amd64"}

### Environment

- **Cloud provider or hardware configuration**:
- **OS** (e.g. from /etc/os-release):
- **Kernel** (e.g. `uname -a`):
- **Install tools**:
- **Others**:

### What happened

I am running `kubectl edit KongIngress/<name>` to edit an existing KongIngress. After making a change like modifying the `route.methods` array, I save the session. However, as I tail the the Kong kubernetes-ingress-controller admin-api logs, I notice that there are no calls to the various admin endpoints that would indicate the configuration is syncing.

### Expected behavior

After editing a KongIngress with `kubectl edit KongIngress/<name>`, the changed configuration is synced with Kong.

### Steps To Reproduce

1. Watch the kubernetes-ingress-controller logs: `kubectl logs -f deploy/kong-ingress-controller admin-api`
2. Edit a KongIngress resource. Be sure to actually change the configuration. `kubectl edit kongingress/<name>`
3. Watch the logs from step 1 and notice that there are no calls to `/upstreams`, `/routes`, etc. endpoints.

**Comment by hbagdi** [comment]
Are you using the DB version or the DB-less version here? It is probably DB, but I want to make sure.

Also, could you please paste here the Ingress and KongIngress resources that you are editing and applying?

**Comment by dsem** [comment]
Yes, using the DB version (postgres).

### KongIngress
```
apiVersion: configuration.konghq.com/v1
kind: KongIngress
metadata:
  name: app-blue
  namespace: prod
proxy:
  connect_timeout: 60000
  path: /api/v1
  protocol: https
  read_timeout: 60000
  retries: 5
  write_timeout: 60000
route:
  methods:
  - GET
  - POST
  - PUT
  - PATCH
  preserve_host: false
  protocols:
  - https
  regex_priority: 0
  strip_path: true
```

### Ingress
```
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  annotations:
    configuration.konghq.com: app-blue
    kubernetes.io/ingress.class: prod-kong
  name: app-blue
  namespace: prod
spec:
  rules:
  - host: <host redacted>
    http:
      paths:
      - backend:
          serviceName: app-blue
          servicePort: 443
        path: /api/v1
      - backend:
          serviceName: app-blue
          servicePort: 443
        path: /ui/v1
status:
  loadBalancer:
    ingress:
    - ip: <ip redacted>
```

**Comment by hbagdi** [comment]
I just tried but I can't reproduce this. Are you watching the Admin API calls of the correct pod? If there are multiple pods then there will be only one Admin API container which will be receiving these requests.

**Comment by dsem** [comment]
Thanks. Yes, I'm sure that I'm watching the correct pods because I only have one pod for the ingress controller. 

I've prepared a minimum working example using minikube that I believe reproduces the issue:

```
# Start from a fresh cluster
minikube delete
minikube start --kubernetes-version v1.13.11

# Set up and postgres and run migrations
kubectl apply -f https://gist.githubusercontent.com/dsem/e306c0084d32172789fec21369f612c6/raw/03409356ac1d0b7ed415d5167926294667a8953c/kong-ingress-controller-issue-474-postgres.yaml
kubectl apply -f https://gist.githubusercontent.com/dsem/c93d2961ef0a901ad6ec6fb8e862fadd/raw/fc560d5f7ed05f16ce0d99f52098bbce40fae2b1/kong-ingress-controller-issue-474-migration.yaml

# Setup Kong, Kong Kubernetes Ingress, and a sample Ingress and KongIngress
kubectl apply -f https://gist.githubusercontent.com/dsem/20a2878cf94feee1108dbd28dcc91152/raw/4e08c28a34328e273773f1b6aacf247064f9b48b/kong-ingress-controller-issue-474-kong-ingress.yaml

# Tail the admin-api logs
kubectl logs -f deploy/kong-ingress-controller admin-api

# Edit a KongIngress rule, like adding or removing an element to the route.methods array
kubectl edit kongingress/app-blue

# Upon saving and closing the edit, notice in the admin-api logs there are no events synching the changed configuration
```

**Comment by hbagdi** [comment]
I set my environment to exactly how you have and tried to reproduce but I couldn't do it.

I added/remoed elements to  the `methods` array and saw `PUT /routes` requests on the Admin API as expected.

**Comment by dsem** [comment]
Weird.

I've uploaded a video illustrating the issue: https://youtu.be/7XCzGcgJ66Q. It sets up minikube the same way as [my previous comment](https://github.com/Kong/kubernetes-ingress-controller/issues/474#issuecomment-562971101). The last ~7 minutes are waiting for ingress controller to sync so you can fast forward until the end. At the end, you can see the ingress controller making admin API commands at it's regular 10-minute interval. Only after that does the added DELETE method show up from the `/routes` endpoint.

**Comment by hbagdi** [comment]
Okay, so from what I can see. A sync is not triggered when you save the KongIngress resource, which should not be happening. I'm stumped as to why this is happening.

Can you try out `kubectl apply -f https://bit.ly/k4k8s` for me?
That is a DB-less version but let's see if that also has that problem or not.

**Comment by dsem** [comment]
Sorry for the delay, I've been away for a couple weeks on vacation for the holidays.

Using your DB-less configuration in minikube, I *am* seeing syncs when the KongIngress is edited.

**Comment by hbagdi** [comment]
I've not been able to reproduce this and I can't find why or how this would happen.

**Comment by guneemwelloeux** [comment]
I'm getting the same sort of behaviour (no immediate Kong update) when I create a new KongConsumer resource. I'm running a DB-less Kong ingress controller 0.7.0

**Comment by hbagdi** [comment]
Are you using a custom Ingress Class?
If so, you need to annotate your KongConsumer resource with `kubernetes.io/ingress.class` annotation as well.

**Comment by guneemwelloeux** [comment]
Yes I am using a custom ingress class. The reason behind that is that I'm using another 2 ingress controllers in the same cluster, and I don't want to have the "no class" ingresses to be kong by default, as is the case right now (which is not really a good thing IMHO.)

The first problem was not the annotation on  the KongConsumer, but it was that consumers cannot be configured in "db-less" mode. I figured that when trying to POST a new consumer on the Kong API directly (port 8444) and got an error message telling me just that.

I haven't been able to test the API key just now, as I'm still struggling setting up the CORS plugin along with that.

**Comment by guneemwelloeux** [comment]
Now the issue I'm having is that updating a plugin configuration (CORS in my case) doesn't trigger the plugin configuration on the individual routes to be updated. I have to edit the ingress itself, and switch the orders of the plugins in my list to force trigger the change to be picked up.

This makes it very unwieldy to debug but also to apply new changes in production (we recently had to add origins, and I don't really see myself having to edit ingresses by hand just so that the ingress controller notices a change)

Is that a feature, or a bug?

**Comment by hbagdi** [comment]
You can't create anything via Admin API in DB-less mode. You need to use CRDs for that.

Also, your problems seem to be different from the original issue (even if the symptoms are same), let's not hijack this issue. Please open a new Github issue if you have questions or issues.

**Comment by hbagdi** [comment]
The fix for this is merged into dev. Stay tuned for 0.10 release.
